src/components/core/ContextSelector/index.tsx

```diff
@@ -1,177 +1,139 @@
+        );
+    }, [contexts, currentContext, queryText, isQuerying]);
 
     React.useEffect(() => {
-        if (currentContext && !types.find(type => currentContext.type.id === type)) {
-            exchangeCurrentContext();
-        }
-    }, [currentContext, types]);
+        search(queryText);
+    }, [queryText]);
 
     const setCurrentContextAsync = async (context?: Context) => {
-        if (context) {
-            await contextManager.setCurrentContextAsync(context);
-            setQueryText('');
-            setRelatedContexts([]);
-            setShouldShowResultsDropdown(false);
-        }
+        if (context) await contextManager.setCurrentContextAsync(context);
     };
 
-    const onSearchChange = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
-        setQueryText(e.target.value);
-    }, []);
-
-    const onOpen = React.useCallback(() => setShouldShowResultsDropdown(true), []);
-    const onClose = React.useCallback(() => {
-        setShouldShowResultsDropdown(false);
-        setQueryText('');
-    }, []);
-
-    const onKeyUp = React.useCallback((e: React.KeyboardEvent<HTMLInputElement>) => {
-        if (e.keyCode === 27) {
-            // ESC
-            onClose();
-        }
-    }, []);
-
-    React.useEffect(() => {
-        if (shouldShowResultsDropdown && inputRef) {
-            inputRef.focus();
-        }
-    }, [shouldShowResultsDropdown, inputRef]);
-
-    const containerRef = React.useRef<HTMLDivElement>(null);
-
-    const isQuerying = useContextSearchResults(
-        inputRef,
-        containerRef,
-        types,
-        relatedContexts,
-        setCurrentContextAsync,
-        queryText,
-        shouldShowResultsDropdown,
-        onClose
-    );
-
-    const inputClassNames = classNames(styles.searchInput, {
-        [styles.focus]: shouldShowResultsDropdown,
+    const dropdownController = useDropdownController((ref, isOpen, setIsOpen) => {
+        const selectedItem = React.useMemo(() => {
+            const mergedItems = mergeDropdownSectionItems(dropdownSections);
+            const selectedItem = mergedItems.find(option => option.isSelected === true);
+            return selectedItem;
+        }, [dropdownSections]);
+
+        const selectedValue = React.useMemo(() => {
+            if (isOpen) {
+                return queryText;
+            } else if (selectedItem) {
+                return selectedItem.title;
+            } else if (currentContext) {
+                return currentContext.title;
+            }
+            return '';
+        }, [isOpen, queryText, selectedItem, currentContext]);
+
+        return (
+            <>
+                <SearchIcon color="#DADADA" />
+                <input
+                    type="text"
+                    value={selectedValue}
+                    onChange={e => setQueryText(e.target.value)}
+                    onClick={() => !isOpen && setIsOpen(true)}

```

Comment:
We should use React.useCallback for these types of handlers to avoid recreating functions on each render and forcing the input component to rerender unnecessarily. We should also have a onKeyUp to catch the escape key and close the selector

Reply:
☝️